### PR TITLE
update ZSON string escaping, delete bstring, simplify zeekio

### LINKED ISSUE #3041: Recognize \u escape sequences in ZSON strings
The ZSON parser does not recognize the `\u` escape sequences for strings [described in the ZSON specification](https://github.com/brimdata/zed/blob/main/docs/formats/zson.md#331-string-escape-rules).

### PR DESCRIPTION
**Pull request by mccanne**
As part of the final updates to the Zed formats, this commit
gets rid of the bstring type and updates the ZSON escaping
rules.  We now convert Zeek TSV strings, with their \x escapes,
into normal UTF-8 Zed strings.  Where an \x escape results in
illegal code points, we escape the the backslash before the x
just as the Zeek does for JSON logs.

Also, with this change it became clear that tzngio was no longer
needed and was consolidated into zeekio.  tableio and textio now
use zeekio.FormatValue().

Closes #3041

### CODE REVIEW

File: zio/zeekio/builder.go [hunk 1]
```diff
@@ -166,11 +167,31 @@ func (b *builder) appendPrimitive(typ zed.Type, val []byte) error {
 		b.buf = zed.AppendBool(b.buf[:0], v)
 	case zed.IDString:
 		// Zeek's enum type is aliased to string.
+		val = unescapeZeekString(val)
+		if !utf8.Valid(val) {
+			// Zeek has an unusual escaping model for non-valid UTF
+			// strings in their JSON integration: invalid bytes are
+			// formatted as the sequence '\' 'x' h h to indicate
```
Comment (mattnibs) [inline zio/zeekio/builder.go:174]:
```suggestion
			// formatted as the sequence '\' 'x' to indicate
```

File: zio/zeekio/builder.go [hunk 2]
```diff
@@ -166,11 +167,31 @@ func (b *builder) appendPrimitive(typ zed.Type, val []byte) error {
...
+			// the presence of unexpected, invalid binary data where
+			// a string was expeceted, e.g., in a field of data coming
+			// of the network.  This is a reasonable scheme; however,
```
Comment (mattnibs) [inline zio/zeekio/builder.go:177]:
```suggestion
			// off the network.  This is a reasonable scheme; however,
```

File: zio/zeekio/escape.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+import (
+	"bytes"
+	"unicode"
+	"unicode/utf8"
+
+	"github.com/brimdata/zed"
+)
+
+// ShouldEscape determines if the given code point at the given position
```
Comment (mattnibs) [inline zio/zeekio/escape.go:11]:
```suggestion
// shouldEscape determines if the given code point at the given position
```

File: zio/zeekio/escape.go [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+// in a value should be escaped for the given output format.  This function
+// does not account for unprintable characters, its main purpose is to
+// centralize the logic about which characters are syntatically significant
+// in each output format and hence must be escaped.  The inContainer parameter
+// specifies whether this values is inside a set or vector (which is needed
+// to correctly implement  zeek log escaping rules).
```
Comment (mattnibs) [inline zio/zeekio/escape.go:17]:
```suggestion
// to correctly implement zeek log escaping rules).
```

File: zson/lexer.go [hunk 5]
```diff
@@ -300,6 +318,123 @@ func (l *Lexer) scanString() (string, error) {
...
+					k += 6
+					if dec := utf16.DecodeRune(r, r2); dec != unicode.ReplacementChar {
+						// A valid pair; consume.
+						if _, err := b.WriteRune(dec); err != nil {
+							return "", err
+						}
+					}
+					//XXX
+					// Invalid surrogate; fall back to replacement rune.
+					//rr = unicode.ReplacementChar
```
Comment (mattnibs) [inline zson/lexer.go:399]: Should this just return an error?
```suggestion
					return "", errors.New("invalid utf16 surrogate pair")
```

**Review by mattnibs — APPROVED** [review]

File: proc/join/ztests/comparable-types.yaml [hunk 6]
```diff
@@ -1,20 +0,0 @@
...
-    data: |
-      {a:"hello",va:"value a"}
-  - name: b.zson
-    data: |
-      {b:"hello"(bstring),vb:"value b"}
-
-outputs:
-  - name: stdout
-    data: |
-      {a:"hello",va:"value a",vb:"value b"}
```
[the remark this answers is not in the document]
Comment (mccanne) [reply proc/join/ztests/comparable-types.yaml:20]: This is covered by the other tests.  The test I took out was making sure join on string=bstring worked too but there is no longer a bstring.

File: zio/zeekio/builder.go (hunk 1 above)
Comment (mccanne) [reply zio/zeekio/builder.go:174]: There is probably a better way to say this but the point is subtle and I think dropping the salient piece of the encoding only makes it more confusing.

File: zson/lexer.go (hunk 5 above)
Comment (mccanne) [reply zson/lexer.go]: Yes, I made an issue.

File: zio/zeekio/escape_test.go [hunk 7]
```diff
@@ -0,0 +1,32 @@
+package zeekio
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+//XXX move back to package zng
+func TestUnescapeBstring(t *testing.T) {
```
Comment (nwt) [inline zio/zeekio/escape_test.go:10]:
```suggestion
func TestUnescapeZeekString(t *testing.T) {
```

File: zio/zeekio/ztests/zeek-utf8-json.yaml [hunk 8]
```diff
@@ -0,0 +1,80 @@
+zed: "*"
+
+# These test inputs are from src/util.cc in the Zeek source code as of Jan 2022.
+# See https://github.com/zeek/zeek/blob/3d9d6e953be5ae8b72336031b6e6659222f52cea/src/util.cc#L2521
+
+# XXX PR question: do we want the funny chars to be u-escaped?  also, our code
```
Comment (nwt) [inline zio/zeekio/ztests/zeek-utf8-json.yaml:6]: If you're asking whether we want to u-escape characters whose Go rune value exceeds `utf.RuneSelf`, then I think the answer is no.

File: zio/zeekio/ztests/zeek-utf8-json.yaml [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+# differs from Zeek on "Private Use Area (E000-F8FF) are always invalid"... "\xee\x8b\xa0
```
Comment (nwt) [inline zio/zeekio/ztests/zeek-utf8-json.yaml:7]: I think this difference is fine.

File: zio/zeekio/ztests/zeek-utf8-json.yaml [hunk 10]
```diff
@@ -0,0 +1,80 @@
+zed: "*"
+
+# These test inputs are from src/util.cc in the Zeek source code as of Jan 2022.
+# See https://github.com/zeek/zeek/blob/3d9d6e953be5ae8b72336031b6e6659222f52cea/src/util.cc#L2521
```
Comment (nwt) [inline zio/zeekio/ztests/zeek-utf8-json.yaml:4]: These are great tests, and I'm happy to have them here if that's your preference, but I do think they'd be more readable if moved to `escape_test.go`.

**Review by nwt — APPROVED** [review]

File: zio/zeekio/ztests/zeek-utf8-json.yaml (hunk 10 above)
Comment (mccanne) [reply zio/zeekio/ztests/zeek-utf8-json.yaml:4]: As discussed, we can move them later or add another column to the Zeek TSV values as a comment explaining each test (taking the comments from the respective tests in the Zeek source code)

**Pull request merged**